### PR TITLE
sampled_addmm: BSR support

### PR DESCRIPTION
**Pull request by nikitaved**
This PR implements a `sampled_addmm` kernel that works with a BSR mask.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101978
* #102095
* __->__ #101163

### CODE REVIEW

**Comment by nikitaved** [comment]
@cpuhrsch , @amjames , this one is ready for review. I will update dispatch in the follow-up. Luckily, we do not have to deal with anything complicated.

File: torch/sparse/_triton_ops.py [hunk 1]
```diff
@@ -88,11 +161,190 @@ def valid_grid_dim(g, mg):
...
+            else:
+                acc_block = alpha * acc_block + beta * tl.load(values_block_ptrs)
+
+            # write result
+            tl.store(values_block_ptrs, acc_block.to(values_ptr.dtype.element_ty))
+
+            # advance val/col_index ptrs to the next block in the row.
+            values_block_ptrs += values_nnz_stride
+            col_index_nnz_ptr += col_indices_stride
+
```
Comment (nikitaved) [inline torch/sparse/_triton_ops.py:347]: @cpuhrsch , @amjames , we can fuse this kernel with `softmax` and `dropout` all with the triton's high-level tools. I suggest we do it here in a follow-up, then I could work on backward.

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -3468,6 +3469,115 @@ def test_triton_bsr_dense_bmm_error_messages(self, device, dtype):
...
+
+                if dtype is torch.float:
+                    res_csr = torch.sparse.sampled_addmm(csr, mat1csr, mat2csr, alpha=alpha, beta=beta)
+                    self.assertEqual(res_tri.to_dense(), res_csr.to_dense())
+
+                # Check grid consistency
+                grid_size = (3,)
+                grid_gen = itertools.product(grid_size, repeat=2)
+                for grid in grid_gen:
+                    res_tri_grid = sampled_addmm(bsr, mat1, mat2, alpha=alpha, beta=beta)
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply test/test_sparse_csr.py:3578]: Oh, right! Just missed it! Good catch, thanks!

File: test/test_sparse_csr.py [hunk 3]
```diff
@@ -3468,6 +3469,125 @@ def test_triton_bsr_dense_bmm_error_messages(self, device, dtype):
...
+        from torch.sparse._triton_ops import sampled_addmm, broadcast_batch_dims_bsr, tile_to_blocksize
+
+        # Note that each value in a non-zero block is in range block_size * [low^2, high^2).
+        tensor = partial(make_tensor, device=device, dtype=dtype, low=0.3, high=1.2)
+
+        # NOTE: batch dims with zero sizes are not supported in `to_sparse_bsr`.
+        batches = [(), (2,), (2, 2)]
+        size = [128, 256, 0]
+
+        def sampled_addmm_ref(input, mat1, mat2, alpha, beta):
```
Comment (cpuhrsch) [inline test/test_sparse_csr.py:3489]: Could another way to implement this be `torch.addmm(input.to_dense(), mat1, mat2, alpha, beta).sparse_mask(input.to_sparse()).to_sparse_bsr(input.values().shape[-2:])`?

Sure it's less efficient, but a simpler reference implementation could give us more confidence?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
I'm a bit worried about the complexity of the reference addmm implementation just so we can have a simple comparison point.

File: test/test_sparse_csr.py [hunk 4]
```diff
@@ -3468,6 +3469,125 @@ def test_triton_bsr_dense_bmm_error_messages(self, device, dtype):
...
+
+                batch_broadcasted_shape = torch.broadcast_shapes(*(t.shape[:-2] for t in (input, mat1, mat2)))
+                self.assertTrue(res_tri.shape == batch_broadcasted_shape + (m, n))
+                self.assertEqual(res_tri, res_ref)
+
+                if dtype is torch.float:
+                    res_csr = torch.sparse.sampled_addmm(csr, mat1csr, mat2csr, alpha=alpha, beta=beta)
+                    self.assertEqual(res_tri.to_dense(), res_csr.to_dense())
+
+                # Check grid consistency
```
Comment (cpuhrsch) [inline test/test_sparse_csr.py:3584]: What does this mean?

File: test/test_sparse_csr.py (hunk 3 above)
Comment (nikitaved) [reply test/test_sparse_csr.py:3489]: We test against CSR from `sparse.sampled_addmm` for more confidence. The proposed solution also only works for non-batched inputs, so, for simplicity, we still need to loop over batches...

File: test/test_sparse_csr.py (hunk 4 above)
Comment (nikitaved) [reply test/test_sparse_csr.py:3584]: I will update the comment to clarify on that.

File: test/test_sparse_csr.py (hunk 3 above)
Comment (cpuhrsch) [reply test/test_sparse_csr.py:3489]: Yes, testing against CSR with float would work too. That's even simpler.

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
Looks like there's one lint error left otherwise this is good to go. Thanks for writing this!

**Pull request closed without merging**